### ISSUE TITLE
Logout  redirect issue.

### ISSUE THREAD

**Issue by zebamba** [IdentityServer/IdentityServer3#2156]
Hi I have posted an issue/question before https://github.com/IdentityServer/IdentityServer3/issues/1968

I did enable logging and found that on the validation https://github.com/IdentityServer/IdentityServer3/blob/master/source/Core/Validation/EndSessionRequestValidator.cs on line 64

<pre>
 if (!subject.Identity.IsAuthenticated && _options.AuthenticationOptions.RequireAuthenticatedUserForSignOutMessage)
{
    Logger.Warn("User is anonymous. Ignoring end session parameters");
    return Invalid();
}
</pre>

I believe that the option it is not set.

Where on the documentation I found how to set "RequireAuthenticatedUserForSignOutMessage"?

Then I can continue troubleshooting my logout issue.

**Comment by zebamba** [comment]
I found here 

https://github.com/IdentityServer/IdentityServer3/blob/bcf35030b67db96ce11c321394b7e3cd84098b73/source/Core/Configuration/AuthenticationOptions.cs

**Comment by zebamba** [comment]
I had to add the source to my project and debug it.

I found in my case that the file https://github.com/IdentityServer/IdentityServer3/blob/01f2446543d8f0c2252d565deb615ef2f3431319/source/Core/Validation/TokenValidator.cs

on line 315 it is not retrieving the client_id

Here is the jwt object
<pre>
{
    "typ": "JWT",
    "alg": "RS256",
    "x5t": "a3rMUgMFv9tPclLa6yF3zAkfquE",
    "kid": "a3rMUgMFv9tPclLa6yF3zAkfquE"
}. {
    "client_id": "MyClientId",
    "scope": "iassist",
    "sub": "a4c6f557-e83d-4e6e-8a67-cb4bbe216f69",
    "amr": ["password"],
    "auth_time": 1447272645,
    "idp": "idsrv",
    "email": "email@email.com",
    "role": ["Administrator", "Role2", "Role3"],
    "name": "User",
    "iss": "https://idsrv3.com",
    "aud": "https://idsrv3.com/resources",
    "exp": 1447273005,
    "nbf": 1447272645
}
</pre>

var clientId = jwt.Audiences.FirstOrDefault();

this line returns the aud not client_id what makes my client not to be valid on line 68

<pre>
if (clientId.IsMissing())
   {
      Logger.Error("No clientId supplied, can't find id in identity token.");
      return Invalid(Constants.ProtectedResourceErrors.
   }
</pre>

Any thoughts?

**Comment by leastprivilege** [comment]
You seem to send an access token. Not the id token.

**Comment by zebamba** [comment]
Yes. 

I am using the access_token.

I am looking to the code that was not implemented by me, but how do I get the Id toked?

Do I need to set on the client to IncludeJwtId?

Do you have an example that I can look to it?

Thank you for all your help.

**Comment by zebamba** [comment]
I found where to modify  my code.

https://identityserver.github.io/Documentation/docsv2/endpoints/authorization.html

My reponse type needs to request the id token

**Comment by zebamba** [comment]
I added the response type as id_token token. There is a space between id_token and token.

I have this error:
<pre>
    error=invalid_request

    The authorization server does not support the requested response type.
</pre>

What am I doing wrong here?

Thank you again

**Comment by zebamba** [comment]
This what I found

this file https://github.com/IdentityServer/IdentityServer3/blob/01f2446543d8f0c2252d565deb615ef2f3431319/source/Core/Validation/AuthorizeRequestValidator.cs line 193

<pre>
if (!Constants.SupportedResponseTypes.Contains(responseType))
    {
        LogError("Response type not supported: " + responseType, request);
        return Invalid(request, ErrorTypes.User, Constants.AuthorizeErrors.UnsupportedResponseType);
            
     }
</pre>

responseType value is "id_token+token"

on this file https://github.com/IdentityServer/IdentityServer3/blob/fc21dad08533b96a6d8672f6bc594dd8694fb2bb/source/Core/Constants.cs line 139

the ResponseTypes has been set as IdTokenToken = "id_token token";

That is why it is returning an error.

Any thoughts?

**Comment by leastprivilege** [comment]
Turn on logging

**Comment by zebamba** [comment]
I got it to work.

You can close the question.

Thank you.

**Comment by BenjaminWang1031** [comment]
@zebamba May I know what was the issue and how did you fix that? I am having the same problem now...

**Comment by zebamba** [comment]
@BenjaminWang1031 

IdentityServer on the application that I had this issue was replaced to another Identity Manager. 

But I remember that on the redirect url I had to change the type of the token and provide the id_token not access token.

something like:  (This was is part of a JavaScript file)

The variable token is the  id_token.

var url = logoutUrl + '?' 
                      + 'id_token_hint=' + encodeURIComponent(token) + '&'
                      + 'post_logout_redirect_uri=' + encodeURIComponent(redirectUrl);                    

Then redirect to that url.            

I am not sure if this can help you, I am sorry this was implemented long time ago and as I stated the IdentityServer as replaced.
